tests/arch/arm/arm_irq_advanced_features/src/arm_irq_target_state.c

```diff
@@ -0,0 +1,77 @@
+/*
+ * Copyright (c) 2020 Nordic Semiconductor ASA.
+ *
+ * SPDX-License-Identifier: Apache-2.0
+ */
+
+#include <ztest.h>
+#include <arch/cpu.h>
+#include <arch/arm/aarch32/cortex_m/cmsis.h>
+
+#if defined(CONFIG_ARM_SECURE_FIRMWARE) && \
+	defined(CONFIG_ARMV7_M_ARMV8_M_MAINLINE)
+
+extern irq_target_state_t irq_target_state_set(unsigned int irq,
+	irq_target_state_t target_state);
+extern int irq_target_state_is_secure(unsigned int irq);
+
+void test_arm_irq_target_state(void)
+{
+	/* Determine an NVIC IRQ line that is not currently in use. */
+	int i;
+
+	for (i = CONFIG_NUM_IRQS - 1; i >= 0; i--) {
+		if (NVIC_GetEnableIRQ(i) == 0) {
+			/*
+			 * Interrupts configured statically with IRQ_CONNECT(.)
+			 * are automatically enabled. NVIC_GetEnableIRQ()
+			 * returning false, here, implies that the IRQ line is
+			 * either not implemented or it is not enabled, thus,
+			 * currently not in use by Zephyr.
+			 */
+
+			/* Set the NVIC line to pending. */
+			NVIC_SetPendingIRQ(i);
+
+			if (NVIC_GetPendingIRQ(i)) {
+				/* If the NVIC line is pending, it is
+				 * guaranteed that it is implemented.
+				 */
+				break;
+			}
+		}
+	}
```

Comment:
I don't really understand this (in my head all logic here should be reversed). What is the difference between "in use", "enabled" and "implemented"? Maybe explain more of why you are looking for an IRQ not in use?

Reply:
So, I want to determine an unused IRQ line to employ for my test.
the logic is as follows:
1. I need an IRQ line that is not in use, currently, so my test won't interfere with Zephyr kernel or peripheral drivers.
2. IRQ lines "in-use" are already enabled during kernel/driver initialization (if not, nothing in this test will enable them so we are, again, safe).
3. We use NVIC_GetEnableIRQ to see if the line is enabled
4. If the line is not enabled we still need to check if the line is wired-in, otherwise it might just not be implemented on the SoC. That's why we pend it and check if it is pended. 
5. Strictly speaking we need to un-pend it, also, and check if it can be un-pended. Then we are 100% sure that the line exists, and can be used for SW interrupt triggering

I'll add the un-pending logic in the test
Sounds ok?

Reply:
Ok, thanks for the explanation!